src/ngraph/runtime/gpu/gpu_emitter.cpp

```diff
@@ -571,9 +571,64 @@ cudnnSetOpTensorDescriptor(opTensorDesc,
                 // Other cases (reordering of axes for tensors with rank>2) are not handled yet.
                 else
                 {
-                    throw runtime_error(
-                        "Axis permutation in reshape is not implemented yet for tensors with "
-                        "rank>2");
+                    std::vector<size_t> input_strides(arg_rank);
+                    std::vector<size_t> output_strides(arg_rank);
+                    std::vector<size_t> trans_strides(arg_rank);
+                    size_t stride = 1;
+                    for (int i = arg_rank - 1; i >= 0; i--)
+                    {
+                        input_strides[i] = stride;
+                        stride *= arg_shape[i];
+                    }
+                    stride = 1;
+                    for (int i = arg_rank - 1; i >= 0; i--)
+                    {
+                        output_strides[i] = stride;
+                        stride *= arg_shape[input_order[i]];
+                    }
+                    for (int i = 0; i < arg_rank; i++)
+                    {
+                        trans_strides[input_order[i]] = output_strides[i];
+                    }
+                    writer << "{   // " << node->get_name() << "\n";
+                    writer.indent++;
+                    writer << "size_t rank = " << arg_rank << ";\n";
+                    writer << "std::vector<size_t> input_strides_h = {" << input_strides[0] << "UL";
+                    for (int i = 1; i < arg_rank; i++)
+                    {
+                        writer << ", " << input_strides[i] << "UL";
+                    }
+                    writer << "};\n";
+
+                    writer << "std::vector<size_t> trans_strides_h = {" << trans_strides[0] << "UL";
+                    for (int i = 1; i < arg_rank; i++)
+                    {
+                        writer << ", " << trans_strides[i] << "UL";
+                    }
+                    writer << "};\n";
+
```

Comment:
Because we are runtime compiling these kernels, ideally we should not have to perform this memcpy. The only reason it is necessary here is because the cuda is emitted from the LLVM compiled function instead of the transformer. If the cuda kernels were emitted at this level, it could be adjusted from within the gpu_emitter to include these vectors/arrays, and then be compiled before the LLVM call is made. 

Last weekend I started a retooling to decouple the cuda kernel emission and execution to enable this kind of functionality -- I haven't followed up on it during the week yet, instead focusing on ops. For now these copies are OK but we should keep a mental note about coming back to this when the compilation order is changed.

Reply:
Absolutely true. This is a hack for now.